### PR TITLE
feat: Add dockerized event webhook consumer example

### LINKED ISSUE #765: Create a Dockerized event webhook consumer
#### Issue Summary

SendGrid has an [Event Webhook](https://sendgrid.com/docs/API_Reference/Event_Webhook/event.html) which posts events related to your email activity to a URL of your choice. We want to have an easily deployable solution that allows for customers to easiy get up and running processing (parse and save) their event webhooks.

#### Acceptance Criteria

* A folder which contains a Dockerized solution
* Example documentation on how to deploy locally and start processing events
* Example documentation on how to deploy on at least one cloud service and start processing events
* An example complete payload with a test to verify the data is parsed properly

### PR DESCRIPTION
**Pull request by KoditkarVedant**
Fixes #765 and #572 

### Checklist
- [x] I have made a material change to the repo (functionality, testing, spelling, grammar)
- [x] I have read the [Contribution Guide] and my PR follows them.
- [x] I updated my branch with the master branch.
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation about the functionality in the appropriate .md file
- [ ] I have added in line documentation to the code I modified

SendGrid has an Event Webhook which posts events related to your email activity to a URL of your choice. We need to have an easily deployable solution that allows for customers to easily get up and running processing (parse and save) their event webhooks.

### CODE REVIEW

**Comment by thinkingserious** [comment]
Hello @KoditkarVedant,<br /><br/>Thanks again for the PR!<br />                                              <br />It's <a href="https://sendgrid.com/blog/hacktoberfest-2018-has-arrived/">HACKTOBERFEST</a>! We want to show our appreciation by sending                                              you some special Hacktoberfest swag. If you have not already, could you please fill out                                              <a href="https://dx.sendgrid.com/hacktoberfest/join">this form</a> so we can send it to you? Thanks!                                             <br /><br />Team SendGrid DX

**Comment by thinkingserious** [comment]
Hello @KoditkarVedant,<br /><br/>Thanks again for the PR!<br /><br />                                         You have earned 30 out of the 5 points needed to receive glorious SendGrid Hacktoberfest swag.<br /><br />                                         Please take a moment to checkout <a href="https://dx.sendgrid.com/hacktoberfest/issues">this link</a> to find more issues to get you past the required threshold or to simply continue the celebration.<br /><br />                                         Also, <strong>please be sure you have officially registered with us <a href="https://dx.sendgrid.com/hacktoberfest/join">here</a> by November 1, 2018</strong> to qualify.<br /><br />                                         If you have any questions you can email us at <a href="mailto:dx+hacktoberfest2018@sendgrid.com">dx+hacktoberfest2018@sendgrid.com</a>.<br /><br />                                         Thank you and Happy Hacktobering!<br /><br />                                         Team SendGrid

**Comment by os1r1s110** [comment]
Hi there, I would just like to say the idea of a containerized EventKit is awesome!
 
Is there a reason why this was never merged in the end? Is there another/better alternative?

**Comment by childish-sambino** [comment]
@os1r1s110 We're (slowly) working through the backlog of open issues/PRs. This one is near the top of that list (note that tests must be passing to be considered at this point).

**Comment by KoditkarVedant** [comment]
@childish-sambino Let me know if there is anything that needs to be done to get this PR merged

File: examples/sendgrid-event-webhook-consumer/Src/EventWebhook/Models/UnsubscribeEvent1.cs [hunk 1]
```diff
@@ -0,0 +1,4 @@
+﻿namespace EventWebhook.Models
```
Comment (childish-sambino) [inline examples/sendgrid-event-webhook-consumer/Src/EventWebhook/Models/UnsubscribeEvent1.cs:1]: Filename ends with `1`.

File: examples/sendgrid-event-webhook-consumer/.gitignore [hunk 2]
```diff
@@ -0,0 +1,330 @@
+## Ignore Visual Studio temporary files, build results, and
```
Comment (childish-sambino) [inline examples/sendgrid-event-webhook-consumer/.gitignore:1]: File is not needed.

File: examples/sendgrid-event-webhook-consumer/LICENSE [hunk 3]
```diff
@@ -0,0 +1,21 @@
+MIT License
```
Comment (childish-sambino) [inline examples/sendgrid-event-webhook-consumer/LICENSE:1]: File is not needed.

File: examples/sendgrid-event-webhook-consumer/README.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+- [Deploy locally](#deploy_locally)
+- [Deploy Heroku](#deploy_heroku)
+- [Testing the Source Code](#testing_the_source_code)
+
+<a name="prerequisite"></a>
+## Prerequisite
+
+Clone the repository
+```
+git clone https://github.com/KoditkarVedant/sendgrid-event-webhook-consumer.git
```
Comment (childish-sambino) [inline examples/sendgrid-event-webhook-consumer/README.md:20]: Multiple references to `github.com/KoditkarVedant`.

File: examples/sendgrid-event-webhook-consumer/README.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
+# sendgrid-events-webhook-consumer
```
Comment (childish-sambino) [inline examples/sendgrid-event-webhook-consumer/README.md:1]: Should move everything to under `examples/eventwebhook/consumer` maybe.

File: examples/sendgrid-event-webhook-consumer/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,21 @@
+FROM microsoft/dotnet:2.1-sdk AS build
+WORKDIR /App
+
+# copy csproj and restore as distinct layers
+COPY *.sln .
+COPY Src/EventWebhook/*.csproj ./Src/EventWebhook/
+COPY Tests/EventWebhook.Tests/*.csproj ./Tests/EventWebhook.Tests/
```
Comment (childish-sambino) [inline examples/sendgrid-event-webhook-consumer/Dockerfile:7]: Can this just copy the whole dir instead of multiple copy commands?

File: examples/sendgrid-event-webhook-consumer/Src/EventWebhook/Views/EventWebhook/Index.cshtml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+﻿
+@{
+    ViewData["Title"] = "Index";
+}
+
+<h1>You have successfully launched the server!</h1>
+<p>
+    Check out <a href="https://github.com/KoditkarVedant/sendgrid-event-webhook-consumer">the documentation</a> on how to use this webhook to utilize the SendGrid Event webhook.
```
Comment (childish-sambino) [inline examples/sendgrid-event-webhook-consumer/Src/EventWebhook/Views/EventWebhook/Index.cshtml:8]: Another `github.com/KoditkarVedant` reference.

File: examples/sendgrid-event-webhook-consumer/Tests/EventWebhook.Tests/EventTests.cs [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+    public class EventTests
+    {
+        [Fact]
+        public async Task AllEvents()
+        {
+            var jsonStream = new MemoryStream(File.ReadAllBytes("TestData/events.json"));
+            
+            IEnumerable<Event> events = await EventParser.ParseAsync(jsonStream);
+
+            events.Count().ShouldBe(11);
```
Comment (childish-sambino) [inline examples/sendgrid-event-webhook-consumer/Tests/EventWebhook.Tests/EventTests.cs:22]: Prefer to test only that there's at least 1 event. Makes it easier to add more events later w/o needing to keep this in-sync.

File: examples/sendgrid-event-webhook-consumer/Dockerfile (hunk 6 above)
Comment (KoditkarVedant) [reply examples/sendgrid-event-webhook-consumer/Dockerfile:7]: @childish-sambino This help us leverage the docker's intermediate layer caching while building the image.

File: examples/sendgrid-event-webhook-consumer/.gitignore (hunk 2 above)
Comment (KoditkarVedant) [reply examples/sendgrid-event-webhook-consumer/.gitignore:1]: done

File: examples/sendgrid-event-webhook-consumer/LICENSE (hunk 3 above)
Comment (KoditkarVedant) [reply examples/sendgrid-event-webhook-consumer/LICENSE:1]: done

File: examples/sendgrid-event-webhook-consumer/README.md (hunk 5 above)
Comment (KoditkarVedant) [reply examples/sendgrid-event-webhook-consumer/README.md:1]: done

File: examples/sendgrid-event-webhook-consumer/Tests/EventWebhook.Tests/EventTests.cs (hunk 8 above)
Comment (KoditkarVedant) [reply examples/sendgrid-event-webhook-consumer/Tests/EventWebhook.Tests/EventTests.cs:22]: Should I remove the model classes too?

Comment (KoditkarVedant) [reply examples/sendgrid-event-webhook-consumer/Tests/EventWebhook.Tests/EventTests.cs]: Ah.. got it. Kept the Processed event as example and removed other events

File: examples/eventwebhook/consumer/Src/EventWebhook/Views/EventWebhook/Index.cshtml [hunk 9]
```diff
@@ -0,0 +1,9 @@
+﻿
+@{
+    ViewData["Title"] = "Index";
+}
+
+<h1>You have successfully launched the server!</h1>
+<p>
+    Check out <a href="https://github.com/sendgrid/sendgrid-csharp/tree/master/examples/eventwebhook/consumer">the documentation</a> on how to use this webhook to utilize the SendGrid Event webhook.
```
Comment (childish-sambino) [inline examples/eventwebhook/consumer/Src/EventWebhook/Views/EventWebhook/Index.cshtml:8]: Switch `master` to `HEAD` as we'll be changing the default branch name soon.

File: examples/sendgrid-event-webhook-consumer/Tests/EventWebhook.Tests/EventTests.cs (hunk 8 above)
Comment (childish-sambino) [reply examples/sendgrid-event-webhook-consumer/Tests/EventWebhook.Tests/EventTests.cs]: Miscommunication here. I'm fine with all the other classes for various event types and think those should be put back, along with the other tests and sample data.

My comment was more about maintenance. Right now there are exactly 11 events in the sample event file. But later I may want to add another one to test out some edge-case parsing or whatever. I'd also have to go into this test file and update the number to 12.

When rather this test doesn't really care exactly _how many_ events were processed; just that they were all processed and there were no exceptions thrown. That's why I think it should just be testing that the deserialized events size is at least 1, since if any of them failed to process then an exception would have been raised.

File: examples/eventwebhook/consumer/Src/EventWebhook/Views/EventWebhook/Index.cshtml (hunk 9 above)
Comment (KoditkarVedant) [reply examples/eventwebhook/consumer/Src/EventWebhook/Views/EventWebhook/Index.cshtml]: done

File: examples/sendgrid-event-webhook-consumer/Tests/EventWebhook.Tests/EventTests.cs (hunk 8 above)
Comment (childish-sambino) [reply examples/sendgrid-event-webhook-consumer/Tests/EventWebhook.Tests/EventTests.cs]: Bump in case this comment was missed.

Comment (KoditkarVedant) [reply examples/sendgrid-event-webhook-consumer/Tests/EventWebhook.Tests/EventTests.cs]: Thanks @childish-sambino, I'd missed that comment - reverted the commit and updated the test as per your suggestion.

**Comment by lvmajor** [comment]
dumb question here, but is this actually doing anything? I read the code quickly, and there doesn't seem to be any persistence layer in the app to save the events somewhere, did I miss something?

**Comment by childish-sambino** [comment]
It has no business logic and is just an example of how to parse and respond to the incoming webhooks. Add business logic here: https://github.com/sendgrid/sendgrid-csharp/blob/11fbf1e385e6c0412962ca4a2d343061f7d696ca/examples/eventwebhook/consumer/Src/EventWebhook/Controllers/EventWebhookController.cs#L30

**Comment by lvmajor** [comment]
Yeah ok that's what I had figured by reading the code. Seems legit, I just thought it would be similar to the old "webhook parser" project hosted on Heroku (based on Rails if my memory's good) which included a persistence layer (Postgres) and a UI to vizualize the events saved. 

Nevertheless, that's a great starting point so that's still a really great addition to the repo :)  

Good job and have a great day!
